ingest/anndata_.py

```diff
@@ -127,3 +133,92 @@ def delocalize_extracted_files(file_path, study_file_id, files_to_delocalize):
                 file,
                 f"_scp_internal/anndata_ingest/{study_file_id}/{file}",
             )
+
+    def process_matrix(self):
+        """Perform matrix processing"""
+        if self.feature_names_unique(self.adata.var_names):
+            self.transform()
+
+    @staticmethod
+    def feature_names_unique(features):
+        """Return True if feature names are unique, else false"""
+        if len(features) == len(features.unique()):
+            return True
+        else:
+            # placeholder for function to provide duplicate feature names
+            # non-happy path - option to use gene IDs?
+            msg = f"Duplicate feature name found"
+            GeneExpression.log_for_mixpanel(
+                "error", "content:duplicate:values-within-file", msg
+            )
+            raise ValueError(msg)
+
+    def transform(self):
+        """Transforms matrix into gene data model."""
+        # initialize settings for mock data loads in tests
+        self.test_models = None
+        self.models_processed = 0
+
+        # derive file name from file path
+        file_name = os.path.basename(self.file_path)
+        start_time = datetime.datetime.now()
+        GeneExpression.dev_logger.info("Starting run at " + str(start_time))
+        num_processed = 0
+        gene_models = []
+        data_arrays = []
+        for all_cell_model in GeneExpression.create_data_arrays(
+            name=f"{file_name} Cells",
+            array_type="cells",
+            values=self.adata.obs.index.tolist(),
+            linear_data_type="Study",
+            linear_data_id=self.study_file_id,
+            **self.data_array_kwargs,
+        ):
+            data_arrays.append(all_cell_model)
+
+        # ASSUMPTION all_cell_model same for raw_count and processed_expression
+        # TODO: if raw counts is indicated check that .raw slot is populated
+
+        # Iterate over feature names (for happy path)
+        for feature in self.adata.var_names.tolist():
+            print(f"processing feature: {feature}")
+            feature_expression_series = sc.get.obs_df(self.adata, keys=feature)
+            if feature_expression_series.hasnans:
+                msg = (
+                    f"Expected numeric expression score - "
+                    f"expression data for \'{feature}\' has NaN values"
+                )
+                GeneExpression.log_for_mixpanel(
```

Comment:
Blocker: this approach seems like it would log to Mixpanel for _each_ NaN feature.  If there are many such features in a given file, that approach could have a few problems, e.g.:

1. Cause ingest to fail with an unhandled error, due to exceptionally many HTTP requests
2. Make the [equivalent CSFV](https://github.com/broadinstitute/single_cell_portal_core/blob/e54fec3c06c68ac49f799b384341d776f6b59311/app/javascript/lib/validation/expression-matrices-validation.js#L358) impossible to compare to this.  CSFV logs one validation error for _all_ NaN violations in a given file; this seems to log one for _each_ such violation.
3. Yield a very noisy set of error messages for the user
4. Flood Mixpanel with noise, substantially depleting quota

So I think we need to rearrange this error logging to report the error once, at the end, like the CSFV linked above.  That seems not especially trivial, but worth doing here, or at least worth doing before the MVP / public beta for AnnData support.

Reply:
The first NaN found will cause an unhandled exception so this approach will only log one NaN error per upload.

Reply:
Ah, thanks for helping me see that -- I no longer consider this blocking.  Perhaps helpful:

```suggestion
                # Logged only once per upload, as first NaN causes unhandled exception
                GeneExpression.log_for_mixpanel(
```